src/python/zquantum/qubo/io.py

```diff
@@ -69,11 +69,32 @@ def load_qubo(input_file: Union[TextIOBase, IO[str], str, PathLike]):
 
 
 def save_qubo(qubo, output_file: Union[TextIOBase, IO[str], str, PathLike]):
```

Comment:
I think we could change it to 
```python 
from _typing import SupportsRead

def save_qubo(qubo, output_file: SupportsRead[Union[str, bytes]]):
    ```

Reply:
I actually can't find any information on `SupportsRead`, the only thing I see is [this](https://mypy.readthedocs.io/en/stable/protocols.html?highlight=SupportsRead#defining-subprotocols-and-subclassing-protocols).
Do you have some better resources?

Reply:
I just looked at the source code of `json.load`
It's in the `_typeshed` module. 
```from _typeshed import SupportsRead
```
likewise `open()`
supports
```from _typeshed import AnyPath
_OpenFile = Union[AnyPath, int]```


But I guess theres already a conflict since json.load and open() does support the same types .... This is why I don't like hinting ... 

Also, PyLance or mypy catches these errors (@brian-dellabetta)

Reply:
So the reason that `open` and `json.load` have different typings is that they actually support *different* types. Namely, `open` supports paths (or files descriptor, hence int in the Union) and `json.load` supports readable objects. In particular `json.load` accepts outputs of `open`. 

In our approach, we are like both `json.dump` and `open`. So we certainly need to support *writeable* things and things that can be used to *create* writeable things. So, using annotations from typeshed it would be something like: `Union[AnyPath. SupportsWrite]`. However we cannot use typeshed in our code, unless we add a dependency to it. It is already installed with mypy but we (corectlly) don't depend on mypy.

So: `str` and `PathLike` take care of `AnyPath` (actually we could alias them like `AnyPath = Union[str, PathLike]`
`IO[str]`  is the same as `TextIO` and takes care of writeable files. However, it is more narrow, so it might be good to copy `SupportsWrite` protocol to our codebase (or define it ourselves).

As a side note, I still think that the approach of json.load/json.dump (i.e. supporting only file like objects instead of patha) is by far the cleanest one... But that's a whole another discussion.

Reply:
Based on what @dexter2206  said I'd leave it as it is, but it's good to know all that – especially when we'll be redesigning the serialization.

Reply:
@dexter2206 awesome thanks for the information! 

It might be nice to define the alias at the top
```Readable= Union[TextIOBase, IO[str], str, PathLike]```or something, just to condense the function definitions and make them more understandable.

Reply:
What about the following:
Let's merge this PR and I'll add such alias to `zquantum.core.utils` and refactor it in all the repos for the sake of clarity and consistency?